subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultTransformerInvoker.java

```diff
@@ -69,41 +73,54 @@
-                                     ClassLoaderHierarchyHasher classLoaderHierarchyHasher) {
+                                     ClassLoaderHierarchyHasher classLoaderHierarchyHasher,
+                                     ProjectFinder projectFinder,
+                                     boolean useTransformationWorkspaces) {
         this.workExecutor = workExecutor;
         this.fileSystemSnapshotter = fileSystemSnapshotter;
         this.artifactTransformListener = artifactTransformListener;
-        this.historyRepository = historyRepository;
+        this.gradleUserHomeHistoryRepository = gradleUserHomeHistoryRepository;
         this.outputFileCollectionFingerprinter = outputFileCollectionFingerprinter;
         this.classLoaderHierarchyHasher = classLoaderHierarchyHasher;
+        this.projectFinder = projectFinder;
+        this.useTransformationWorkspaces = useTransformationWorkspaces;
     }
 
     @Override
-    public boolean hasCachedResult(File primaryInput, Transformer transformer) {
-        return historyRepository.hasCachedResult(getImmutableTransformationIdentity(primaryInput, transformer));
+    public boolean hasCachedResult(File primaryInput, Transformer transformer, TransformationSubject subject) {
+        Optional<ProjectInternal> producerProject = determineProducerProject(subject);
+        return determineHistoryRepository(producerProject).hasCachedResult(getTransformationIdentity(producerProject.isPresent(), primaryInput, transformer, subject));
     }
 
     @Override
     public Try<ImmutableList<File>> invoke(Transformer transformer, File primaryInput, TransformationSubject subject) {
-        TransformationIdentity identity = getImmutableTransformationIdentity(primaryInput, transformer);
-        return historyRepository.withWorkspace(identity, (identityString, workspace) -> {
-            return fireTransformListeners(transformer, subject, () -> {
-                CurrentFileCollectionFingerprint primaryInputFingerprint = DefaultCurrentFileCollectionFingerprint.from(ImmutableList.of(fileSystemSnapshotter.snapshot(primaryInput)), AbsolutePathFingerprintingStrategy.INCLUDE_MISSING);
-                TransformerExecution execution = new TransformerExecution(primaryInput, transformer, workspace, identityString, historyRepository, primaryInputFingerprint, subject.getDependencies());
-                UpToDateResult outcome = workExecutor.execute(execution);
-                return execution.getResult(outcome);
+        return Try.ofFailable(() -> {
+            Optional<ProjectInternal> producerProject = determineProducerProject(subject);
+            TransformerExecutionHistoryRepository historyRepository = determineHistoryRepository(producerProject);
+            TransformationIdentity identity = getTransformationIdentity(producerProject.isPresent(), primaryInput, transformer, subject);
+            return historyRepository.withWorkspace(identity, (identityString, workspace) -> {
```

Comment:
It still feels weird to me that it's the history repository that runs things through a workspace. Seems to me that the workspace provider could get the history repository instead, and maybe that way we'd get something more logical.

Reply:
We have two things:
- workspace provider for the transformation
- execution history for the transformation

Those two things need to be compatible, so both should be from Gradle user home scope, or from project scope and Gradle scope. The thing which ties both from the same scope together is currently the `TransformerExecutionHistoryRepository `. Maybe we should call this class differently?

Reply:
I guess that's what I'm driving at, yes. We should have a thing that ties these things together, have a mutable and an immutable variant of it, and use each in the right scope.

Reply:
I changed it so that `TransformationWorkspaceProvider` ties the things together.